Jcli make transaction

danielSanchezQ opened the pull request:
I'm starting this as a draft. It is a close example of what jcli path to being a lib should lead to. 

The simplified transaction command uses several of the other transaction commands.
I started extracting the common code into external functions. Also stripping some functionality like reading and writing to files that shouldn't be within the reusable computations.

danielSanchezQ commented:
This is related to #3216

enzoc4 commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,68 @@
+pub struct SimplifiedTransaction {
+    /// the account to debit the funds from
+    #[structopt(name = "ACCOUNT")]
+    pub faucet_address: interfaces::Address,
```
I don't really see anything specific to faucets here. Why not just `sender_account` or something?

danielSanchezQ replied:
It was called `faucet_address` in the original script, so I kept it like that. But you are right, gonna rename it :)

eugene-babichenko replied:
Don't pay too much attention to the original script naming. `jcli` is universal and does not care about the details of specific deployments.

enzoc4 commented on jcli/src/jcli_lib/transaction/mod.rs:
```diff
@@ -62,6 +63,8 @@ pub enum Transaction {
     /// get the message format out of a sealed transaction
     ToMessage(common::CommonTransaction),
+    /// send  a transaction from faucet (simplified method)
+    Send(simplified::SimplifiedTransaction),
```
also, a command named `Send` may be a bit confusing, at least to me it suggests network interaction

danielSanchezQ replied:
Makes sense

danielSanchezQ commented on jcli/src/jcli_lib/address.rs:
```diff
@@ -131,7 +131,7 @@ fn mk_delegation(prefix: &str, s: PublicKey<Ed25519>, testing: bool, d: PublicKe
 }
 
-fn mk_account(prefix: &str, s: PublicKey<Ed25519>, testing: bool) {
+pub fn mk_account(prefix: &str, s: PublicKey<Ed25519>, testing: bool) {
```
I have to reverse this, since finally I am not really using it outside.

danielSanchezQ replied:
yeah, sorry, Sometimes I just make annotations so I do not forget to do so 😅

eugene-babichenko replied:
We can make GitHub Actions fail upon meeting TODO/FIXME comments. ;)

danielSanchezQ replied:
It would not work. because it was not even noted as a fixme or todo. The thing is that I usually make a self review, if I spot something that shouldn't be there for any reason I make a note in the review so I do not forget to change it 😄

danielSanchezQ commented on jcli/src/jcli_lib/transaction/mod.rs:
```diff
@@ -62,6 +64,8 @@ pub enum Transaction {
     /// get the message format out of a sealed transaction
     ToMessage(common::CommonTransaction),
+    /// send a transaction from faucet (simplified method)
+    Simplified(simplified::SimplifiedTransaction),
```
I didn't really know how to name this (shortly). Any suggestion is welcome.

eugene-babichenko commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,161 @@
+    // get transaction and block0 ids
+    let transaction_sign_data_hash = transaction.transaction_sign_data_hash();
+    let block0 = load_block(open_block_file(&Some(block0_file))?)?;
+    let block0_hash = block0.header.id();
```
If the only point of using `block0` is to take its hash, then just use its hash as a command argument.

danielSanchezQ replied:
Cool. Just tried to stick to the original script here (and in other places).

danielSanchezQ commented on jcli/src/jcli_lib/rest/config.rs:
```diff
@@ -7,12 +7,12 @@ use std::path::PathBuf;
     /// E.g. `-h http://127.0.0.1`, `--host https://node.com:8443/cardano/api`
     #[structopt(short, long, env = "JORMUNGANDR_RESTAPI_URL")]
-    host: Url,
+    pub host: Url,
```
I needed to access that field that otherwise is just for internal use. If embedded in some other command (as happens here) it is handy to access it. Do you think a method would be better?

eugene-babichenko replied:
> If embedded in some other command (as happens here) it is handy to access it.

Fine if that's the case :+1:

danielSanchezQ commented on jcli/src/jcli_lib/rest/v0/account/mod.rs:
```diff
@@ -24,13 +25,20 @@ impl Account {
+    args: RestArgs,
+    account_id: AccountId,
+) -> Result<AccountState, Error> {
+    Ok(args
```
Actually, yes. Thanks.

danielSanchezQ commented on jcli/src/jcli_lib/transaction/mod.rs:
```diff
@@ -62,6 +64,8 @@ pub enum Transaction {
     Auth(auth::Auth),
     /// get the message format out of a sealed transaction
     ToMessage(common::CommonTransaction),
+    /// send a transaction from faucet (simplified method)
```
I guess you wanted to say it shouldn't?

eugene-babichenko replied:
Yes :sweat_smile:

danielSanchezQ commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,161 @@
+    Ok((sk, address))
+}
+
+#[allow(clippy::too_many_arguments)]
```
It has indeed but is what is needed. Even if splitting it would have to receive such data. I'll try to do something to reduce that anyway.

danielSanchezQ replied:
I could refactor things inside in some smaller sections maybe. But the grouping function would still need to receive that. At the end we need all we have in the cli input to compute :/

eugene-babichenko reviewed (changes requested):
`grep`-ing (or `rg`-ing for fellow Rustaceans) still shows notions of faucet and "simplified transactions".

eugene-babichenko commented:
Definitely not how errors should be reported.

```
thread 'main' panicked at 'Found positional argument which is not required with a lower index than a required positional argument: "change" index 4', /home/babichenko/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-2.33.3/src/app/parser.rs:643:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

eugene-babichenko commented:
The help itself is fine and accessible via `jcli help transaction make-transaction`. This must be an issue with the library.

danielSanchezQ commented:
The failing test looks no related.

eugene-babichenko commented:
It works :+1:. Here are my concerns:

- It outputs `Private key of receiver (to revert transaction for testing purposes)` and I would rather remove it for security reasons.
- We do actual transaction sending here, but (see the word "offline")
```
	❯ jcli transaction --help
	jcli-transaction 0.11.1
	Build and view offline transaction
```
- I think it might be better to set the default values of fees to the Jormungandr defaults.
- The help is still sort of broken (some arguments not showing) and I think this is due to structure fields ordering again.
```
	❯ jcli transaction make-transaction -h
	error: The following required arguments were not provided:
    	<ACCOUNT>
    	<VALUE>
    	<block0-hash>
	
	USAGE:
```

danielSanchezQ commented:
@eugene-babichenko how about?:

> It outputs Private key of receiver (to revert transaction for testing purposes) and I would rather remove it for security reasons.

Maybe add flag to enable disable it?

> We do actual transaction sending here, but (see the word "offline")

Didn't understand this. Could you elaborate?

> I think it might be better to set the default values of fees to the Jormungandr defaults.

Definetly. Can you point me to the values please?

> The help is still sort of broken (some arguments not showing) and I think this is due to structure fields ordering again.

wtf...it actually works with `--help` 😞

eugene-babichenko commented:
> Maybe add flag to enable disable it?

Just remove IMO.

> Didn't understand this. Could you elaborate?

Everything under `jcli transactions` is dedicated to building transactions and do not contact HTTP endpoints.

> Definetly. Can you point me to the values please?

Uhm, well, I guess we don't have defaults for them, or I still did not wake up. Will need to ask.

eugene-babichenko commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,145 @@
+    pub fee: common::CommonFees,
+
+    #[structopt(flatten)]
+    rest_args: RestArgs,
```
They need to be removed since we don't do transaction sending here anymore.

danielSanchezQ replied:
It is still needed to query some account data.

eugene-babichenko replied:
Oopsie, sorry.

danielSanchezQ commented:
Added request fees from rest api

eugene-babichenko commented on testing/jormungandr-integration-tests/src/jcli/transaction/simplified.rs:
```diff
@@ -0,0 +1,50 @@
+        block0_hash.to_string(),
+        sk_file_path,
+        staging_file,
+    );
```
Shall we also make sure the transaction was successfully casted?

danielSanchezQ replied:
It is internally asserted in the `make_transaction` method.

eugene-babichenko commented on testing/jormungandr-integration-tests/src/common/jcli/command/transaction.rs:
```diff
@@ -186,6 +186,30 @@ impl TransactionCommand {
     }
 
+    pub fn make_transaction(
+        mut self,
+        host: String,
+        sender: jormungandr_lib::interfaces::Address,
+        value: jormungandr_lib::interfaces::Value,
+        block0_hash: String,
+        secret: impl AsRef<Path>,
+        staging_file: impl AsRef<Path>,
```
Uhm, but who is the receiver?

danielSanchezQ replied:
The receiver is a testing account. Actually yesterday I was thinking that I may have missed making it optional. So if no address is provided a testing account is used.

eugene-babichenko commented on testing/jormungandr-integration-tests/src/common/jcli/command/transaction.rs:
```diff
@@ -204,9 +205,14 @@ impl TransactionCommand {
+
+        if let Some(receiver) = receiver {
+            self.command
+                .arg("--receiver-account")
```
Maybe just `--receiver`?

eugene-babichenko commented:
1. This question `Continue? Yes[y] or No[n]?` is always returning `canceled by user` regardless of my answer.
2. I cannot post the resulting message with `jcli rest v0 message post`. Looks like we use inconsistent encodings.

danielSanchezQ commented:
> 1. This question `Continue? Yes[y] or No[n]?` is always returning `canceled by user` regardless of my answer.
> 2. I cannot post the resulting message with `jcli rest v0 message post`. Looks like we use inconsistent encodings.

Yep, sorry, I realize late. Was working on a fix atm.

eugene-babichenko reviewed (changes requested):
`jcli rest v0 message post` expects a hex-encoded fragment. Please add a full e2e test that would include sending so that we are sure everything actually works.

danielSanchezQ commented:
@eugene-babichenko I added some flag to directly post the fragment (since we already have the rest data). Also I added the test. If needed the hex encoded one it can always be done through the `to-message` command afterwards.

eugene-babichenko commented:
It worked perfectly, but needs some work on basic usability :)
I will walk through the code and leave my comments later.

eugene-babichenko commented on jcli/src/jcli_lib/utils/io.rs:
```diff
@@ -65,3 +65,12 @@ pub fn read_yaml<D: DeserializeOwned>(path: &Option<impl AsRef<Path>>) -> Result
+
+pub fn ask_yes_or_no(with_output: bool) -> std::io::Result<bool> {
+    if with_output {
+        println!("Continue? Yes[y] or No[n]?");
```
```suggestion
        print!("Continue? Yes[y] or No[n]? ");
```

eugene-babichenko commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,218 @@
+    // ask for user confirmation after adding fees
+    if !force {
+        println!(
+            "Total value to transfer (including fees): {}\n",
```
```suggestion
            "Total value to transfer (including fees): {}",
```

danielSanchezQ replied:
Yep, leftover from a previous message.

eugene-babichenko commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,218 @@
+    // ask for user confirmation after adding fees
+    if !force {
+        println!(
+            "Total value to transfer (including fees): {}\n",
```
```suggestion
            "Total value to transfer (including fees): {}",
```

eugene-babichenko commented on jcli/src/jcli_lib/rest/v0/message/mod.rs:
```diff
@@ -57,14 +57,19 @@ fn get_logs(args: RestArgs, output_format: OutputFormat) -> Result<(), Error> {
+    let fragment =
         Fragment::deserialize(msg_bin.as_slice()).map_err(Error::InputFragmentMalformed)?;
+    let fragment_id = post_fragment(args, fragment)?;
+    println!("{}", fragment_id);
```
This needs to be replaced with a more informative message.

danielSanchezQ replied:
I kept it as it was. I understand that this may be used in order to pipe the output to something else. But I'm down to change it.

eugene-babichenko commented on jcli/src/jcli_lib/transaction/simplified.rs:
```diff
@@ -0,0 +1,218 @@
+        if self.post {
+            let fragment = transaction.fragment()?;
+            let fragment_id = post_fragment(self.rest_args, fragment)?;
+            println!("{}", fragment_id);
```
Same about adding more information here: right now we are just printing numbers and letters.

danielSanchezQ replied:
Likewise here, but it is clearer here that we could output something better, since this is more a one-shot interactive thingy :)

eugene-babichenko reviewed (commented):
We are outputting the raw fragment body somewhere regardless of settings provided by the user. This is not something desirable when, for example, we are using the `--post` flag.

On top of that, since this is a raw body there is no way it could be properly copied to a file. A good approach would be to either output this only to a file or output everything besides the fragment body to `stderr` to allow proper redirects. IMO the first option is more viable in this case (since this is an interactive utility) but I am open to a discussion on that.
